cgen error for `assert foption() == none`

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.6 c689f80.549e11b
OS: linux, Ubuntu 20.04.6 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz

getwd: /space/v/oo
vexe: /space/v/oo/v
vexe mtime: 2024-06-23 06:53:46

vroot: OK, value: /space/v/oo
VMODULES: OK, value: /home/delian/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.43.2
Git vroot status: weekly.2024.25-9-g549e11bf-dirty (2 commit(s) behind V master)
.git/config present: true

CC version: cc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
thirdparty/tcc status: thirdparty-linux-amd64 40e5cbb5

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg a.v`
```v
fn find_token_by_id() ?int {
	return none
}

// This works:
// token := find_token_by_id()
// assert token == none

// This should work too:
assert find_token_by_id() == none
println('done')

```

**What did you expect to see?**

a compiled program

**What did you see instead?**
```
==================
/tmp/v_1000/a.01J11ZAXP67X127N8815H6T592.tmp.c:13156: error: '_t1' undeclared
...
==================
(Use `v -cg` to print the entire error message)

```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2048,9 +2048,11 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 					((expr is ast.SelectorExpr || (expr is ast.Ident && !expr.is_auto_heap()))
 					&& ret_typ.is_ptr() && expr_typ.is_ptr() && expr_typ.has_flag(.option))
 					|| (expr_typ == ret_typ && !(expr_typ.has_option_or_result()
-					&& (expr_typ.is_ptr() || g.table.type_kind(ret_typ) == .placeholder)))
+					&& (expr_typ.is_ptr() || expr is ast.LambdaExpr)))
 				// option ptr assignment simplification
 				if simple_assign {
+					println(g.table.type_kind(ret_typ))
+					println(expr)
 					g.write('${tmp_var} = ')
 				} else if expr_typ.has_flag(.option) && expr is ast.PrefixExpr
 					&& expr.right is ast.StructInit

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2051,8 +2051,6 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 					&& (expr_typ.is_ptr() || expr is ast.LambdaExpr)))
 				// option ptr assignment simplification
 				if simple_assign {
-					println(g.table.type_kind(ret_typ))
-					println(expr)
 					g.write('${tmp_var} = ')
 				} else if expr_typ.has_flag(.option) && expr is ast.PrefixExpr
 					&& expr.right is ast.StructInit

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2045,8 +2045,9 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 				}
 			} else {
 				simple_assign =
-					(expr is ast.SelectorExpr || (expr is ast.Ident && !expr.is_auto_heap()))
-					&& ret_typ.is_ptr() && expr_typ.is_ptr() && expr_typ.has_flag(.option)
+					((expr is ast.SelectorExpr || (expr is ast.Ident && !expr.is_auto_heap()))
+					&& ret_typ.is_ptr() && expr_typ.is_ptr() && expr_typ.has_flag(.option))
+					|| expr_typ == ret_typ
 				// option ptr assignment simplification
 				if simple_assign {
 					g.write('${tmp_var} = ')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2047,8 +2047,8 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 				simple_assign =
 					((expr is ast.SelectorExpr || (expr is ast.Ident && !expr.is_auto_heap()))
 					&& ret_typ.is_ptr() && expr_typ.is_ptr() && expr_typ.has_flag(.option))
-					|| (expr_typ == ret_typ && !(expr_typ.is_ptr()
-					&& expr_typ.has_option_or_result()))
+					|| (expr_typ == ret_typ && !(expr_typ.has_option_or_result()
+					&& (expr_typ.is_ptr() || g.table.type_kind(ret_typ) == .placeholder)))
 				// option ptr assignment simplification
 				if simple_assign {
 					g.write('${tmp_var} = ')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2047,7 +2047,8 @@ fn (mut g Gen) expr_with_tmp_var(expr ast.Expr, expr_typ ast.Type, ret_typ ast.T
 				simple_assign =
 					((expr is ast.SelectorExpr || (expr is ast.Ident && !expr.is_auto_heap()))
 					&& ret_typ.is_ptr() && expr_typ.is_ptr() && expr_typ.has_flag(.option))
-					|| expr_typ == ret_typ
+					|| (expr_typ == ret_typ && !(expr_typ.is_ptr()
+					&& expr_typ.has_option_or_result()))
 				// option ptr assignment simplification
 				if simple_assign {
 					g.write('${tmp_var} = ')

--- a/vlib/v/gen/c/ctempvars.v
+++ b/vlib/v/gen/c/ctempvars.v
@@ -22,4 +22,5 @@ fn (mut g Gen) gen_ctemp_var(tvar ast.CTempVar) {
 	g.write('${styp} ${tvar.name} = ')
 	g.expr(tvar.orig)
 	g.writeln(';')
+	g.set_current_pos_as_last_stmt_pos()
 }

--- a/vlib/v/tests/assert_fn_ret_option_test.v
+++ b/vlib/v/tests/assert_fn_ret_option_test.v
@@ -1,0 +1,8 @@
+fn find_token_by_id() ?int {
+	return none
+}
+
+fn test_main() {
+	assert find_token_by_id() == none
+	println('done')
+}

